runtime/lua/vim/lsp/inlay_hint.lua

```diff
@@ -4,13 +4,32 @@ local ms = require('vim.lsp.protocol').Methods
 local api = vim.api
 local M = {}
 
----@class (private) vim.lsp.inlay_hint.bufstate
+---@class (private) vim.lsp.inlay_hint.globalstate Global state for inlay hints
+---@field enabled boolean Whether inlay hints are enabled for this scope
+---@type vim.lsp.inlay_hint.globalstate
+local globalstate = {
+  enabled = false,
+}
+
+---@class (private) vim.lsp.inlay_hint.bufstate: vim.lsp.inlay_hint.globalstate Buffer local state for inlay hints
 ---@field version? integer
 ---@field client_hints? table<integer, table<integer, lsp.InlayHint[]>> client_id -> (lnum -> hints)
 ---@field applied table<integer, integer> Last version of hints applied to this line
----@field enabled boolean Whether inlay hints are enabled for this buffer
 ---@type table<integer, vim.lsp.inlay_hint.bufstate>
-local bufstates = {}
+local bufstates = vim.defaulttable(function(_)
+  local bufstate = { applied = {} }
+  setmetatable(bufstate, {
+    __index = globalstate,
+    __newindex = function(state, key, value)
+      if globalstate[key] == value then
+        rawset(state, key, nil)
```

Comment:
Is this not problematic? If you explicitly bufstate to a value equal to globalstate, and then change the value, it will change the value for the buffer even though it already has a local value

Reply:
There doesn't seem to be a clear convention for what to do in this situation, so I deliberately designed it like this.

>If you explicitly bufstate to a value equal to globalstate, and then change the value of globalstate, it will change the value for the buffer even though it already is supposed to have a local value

So, this is what I want to achieve.

If we expose variables to users, the situation is simple, as long as the local variable exists, we use the local variable; otherwise we use the global variable. This is very similar to what you expected.

But if the API we exposed is functions like `enable`, It’s hard to say whether the above approach is correct or not. Because there is no way to remove "local variables" by using `enable`. Once the user manages a buffer's `bufstate`, it means that they must manage it all the time, there's no way to make them follow the `globalstate`.

The logic behind this design is that if it can be globally `enable` or locally `enable`, when the user selects local `enable`, we think the user's purpose is to distinguish the local state from the global state. I think this design is good for interaction,

You can think about whether what you said is similar to what I said. When the user explicitly sets the previously unequal `bufstate` to be equal to `globalstate`, he is probably thinking "this buffer is no longer special" rather than "no matter what the `globalstate` is, this `bufstate` should be like this."

Reply:
I'm still not fully convinced. But I'll refer to @justinmk as he's the best person to ask for this kind of thing

Reply:
buffer-local should always override global. the global should act as the "default" in the absence of a buffer-local setting. this is well-established by e.g. buffer-local options.

it sounds like that's what's implemented here, correct?